src/TaggingServiceProvider.php

```diff
@@ -0,0 +1,121 @@
+ * All such "tag" services are empty list in the beginning of their resolution, so it doesn't matter
+ * if it gets overwritten by another module's identical empty list.
+ *
+ * @psalm-import-type Factory from ServiceProvider
+ * @psalm-import-type Extension from ServiceProvider
+ */
+class TaggingServiceProvider implements ServiceProviderInterface
+{
+    /** @var array<Factory> */
+    protected array $factories;
+    /** @var array<Extension> */
+    protected array $extensions;
+
+    public function __construct(ServiceProviderInterface $inner)
+    {
+        $this->factories = $inner->getFactories();
+        $this->extensions = $inner->getExtensions();
+        $this->indexTags();
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getFactories()
+    {
+        return $this->factories;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getExtensions()
+    {
+        return $this->extensions;
+    }
+
+    /**
+     * Indexes tagged factories, and creates factories and extensions for tags.
+     *
+     * @throws ReflectionException If problem obtaining factory reflection.
+     */
+    protected function indexTags(): void
+    {
+        $tags = [];
+
+        foreach ($this->factories as $serviceName => $factory) {
+            if (is_string($factory)) {
+                continue;
+            }
+
+            $reflection = is_object($factory) && get_class($factory) === 'Closure'
+                ? new ReflectionFunction($factory)
+                : new ReflectionObject($factory);
+            $docBlock = $reflection->getDocComment();
+
+            // No docblock
+            if ($docBlock === false) {
+                continue;
+            }
+
+            $factoryTags = $this->getTagsFromDocBlock($docBlock);
+            foreach ($factoryTags as $tag) {
+                if (!isset($tags[$tag]) || !is_array($tags[$tag])) {
+                    $tags[$tag] = [];
+                }
+                $tags[$tag][] = $serviceName;
+            }
+        }
+
+        foreach ($tags as $tag => $taggedServiceNames) {
+            $this->factories[$tag] = fn (): array => [];
+            $this->extensions[$tag] = function (ContainerInterface $c, array $prev) use ($taggedServiceNames): array {
+                return array_merge(
+                    $prev,
+                    array_map(fn (string $serviceName) => $c->get($serviceName), $taggedServiceNames)
+                );
+            };
+        }
+    }
+
+    /**
+     * Retrieves tags names that are part of a docBlock.
+     *
+     * @link https://www.php.net/manual/en/reflectionclass.getdoccomment.php#118606
+     *
+     * @param string $docBlock The docBlock.
+     *
+     * @return array<string> A list of tag names.
+     */
+    protected function getTagsFromDocBlock(string $docBlock): array
+    {
+        $regex = '#(@tag\s*(?P<tags>[^\s]+))#';
```

Comment:
This regex will catch `@tags` that aren't in the "beginning" of the line, such as:

```php
/** There is a misplaced @tag in this comment. */

/**
 * There is a misplaced @tag in this comment.
 */
```

I added to services to the test to confirm this, `serviceY` and `serviceZ` for the above two scenarios, and both were detected.

Might I suggest this regex instead:

```suggestion
        $regex = '#^\s*/?\**\s*(@tag\s*(?P<tags>[^\s]+))#m';
```

I could also commit and push my local changes, if you want. Or submit a PR targeting this branch.

Reply:
If it's possible - please feel free to add the rest of your changes e.g. the test to this very branch directly.